### PR TITLE
ci: add qualified v4 release promotion

### PR DESCRIPTION
**Pull request by dsarno**
## Status

This PR lands release preparation and updater fixes; **it does not authorize or complete v4 publication**. Qualification refuses before final signing while unchanged-candidate runtime, complete failpoint/crash-recovery, and reviewed numeric stress qualification remain incomplete. GLES texture shutdown diagnostics also remain under investigation. The current evidence and remaining gates are recorded in [docs/v4-release-review-followup.md](https://github.com/hi-godot/godot-ai/blob/2589d411523bc82bd54273ebb3d01c90bd6da3f8/docs/v4-release-review-followup.md).

## Summary

- Add immutable A/B packaging and qualification scaffolding, with exact-inventory/signature checks, retained dependencies, offline wheel/sdist tests, and actual documented installs. Incomplete runtime/failure/stress producers cannot pass the complete-evidence gate.
- Promote only independently approved, digest-bound candidate bytes with major/minor/patch selection, main-only human-approved signing/publishing environments, environment-bound PyPI trusted publishing, and verified resumable partial publication.
- Retire the unused beta issue-closure workflow and beta CI triggers.
- Fix built-in handler quiescence and shutdown ownership; busy or unknown handlers still refuse unsafe hot replacement.
- Fix the final-v3 restart handoff and CI launch behavior for #957: persist next-start enablement without loading v4 in the old process, preserve headless drivers across restart, scope bridge failure status to its project, and wait through transient predecessor uncertainty without transferring the lease until positive closure proof. No additional timeout increase.
- Add real-engine restart/failure regressions, bounded progress output, replacement-process diagnostics, and typed reload-window-only transport classification in the stress harness.

## Validation

- Full local Python suite with Godot integration: **2,448 passed, four platform-inapplicable skips**; one existing third-party deprecation warning.
- Full live Godot 4.7.2 suite: **2,150 passed, zero failed, 11 conditional/platform skips, all 71 suites**.
- Fresh owner-operated Update after real tool use: matching plugin/backend `4.0.1`, durable migration, retained backup, no early vNext exit hook and no new crash report.
- Clean exploratory stress: **1,223 explicitly pinned calls, zero unexpected errors, both reloads survived**. This is not locked-profile release approval.
- Full Ruff scope, actionlint, all eleven architecture gates, and `git diff --check` pass.
- [Cross-platform CI for 2589d41](https://github.com/hi-godot/godot-ai/actions/runs/33715456363): all three expanded Godot upgrade-regression steps pass. Windows passed all 19 upgrade/restart/failure cases; its formerly timing-out final-v3 case completed in about 27 seconds. See the run for the complete matrix result.
- All ten existing inline review threads are resolved. The eight-point review's full artifact-qualification requirement remains an explicit release gate, not a claim that this PR is ready to publish v4.

## Summary by CodeRabbit

* **Release Management**
  * Added qualification workflows to build, sign, and verify release candidates across Linux, macOS, and Windows without publishing.
  * Enabled approval-gated promotion of qualified candidates to PyPI and GitHub Releases with verified artifacts.
  * Removed automatic issue-closing behavior after beta merges.

* **Reliability**
  * Improved plugin update and shutdown handling by preventing reloads while background work is still active.
  * Improved session-pinned testing and editor reload stability.

* **Documentation**
  * Updated release and packaging guidance with stricter qualification and publication requirements.

* **CI**
  * Limited continuous integration triggers to the `main` branch.

### CODE REVIEW

**Comment by dsarno** [comment]
Review (posted as a comment — GitHub refuses a formal request-changes on one's own PR):

Diagnosed against `main` at c6ba797. Nothing here is exploitable by an outsider today (`workflow_dispatch` needs write access and nothing has been published), but as written the pipeline cannot run end-to-end, and the two boundaries the docs now claim — "promotion consumes a qualified run" and "the attestation channel is independent" — are not enforced by the workflow. All fixes are small; a second revision should be a quick fix-and-merge. Ranked by security impact.

**Boundaries not enforced**

1. **`release.yml:48-61` — qualification does not gate promotion.** `run-id: ${{ inputs.qualification_run_id }}` downloads `v4-candidate-a/b` from *any* run in the repo; nothing checks the run's `path`, `event`, `head_branch`, or `conclusion`, and the `qualify` job's result is never consulted (its `v4-qualification-<os>` artifacts are dead uploads; `v4-candidate-*` are produced by `build`, before `qualify` runs). A red macOS/Windows row still promotes. And since `ci.yml` runs on `pull_request` from forks, a fork PR can upload artifacts named `v4-candidate-a/b` into a base-repo run. Fix: before any download, `gh api repos/$R/actions/runs/$ID` and assert `.path == ".github/workflows/release-qualification.yml"`, `.event == "workflow_dispatch"`, `.head_branch == "main"`, `.conclusion == "success"`, plus all `/jobs` succeeded; bind `qualification_run_id` into the approval JSON so the attestation names the run, not just digests.
2. **`release.yml:21-25, 69-70` — `approval_repository` is a free-form input** with `dsarno/godot-ai-release-attestations` only as the default, so the dispatcher chooses where the approval comes from. The write-access side of the boundary is honored (only `github.token`); the read side is unpinned. Fix: drop the input, hardcode the repo in `env:`, regex-validate `approval_path` (`^[A-Za-z0-9._/-]+\.json$`, no `..`).
3. **`release-qualification.yml:96` (also `:44-49`, `:79-81`) — expression injection in the signing step.** `--tag "v${{ matrix.version }}" --version "${{ matrix.version }}"` is interpolated in the step whose env holds `RELEASE_SIGNING_KEY_PEM`; `version_a/b` are never validated (only `source_*` get the 40-hex regex). A version like `4.1.0$(curl -d @"$key_file" …)` runs at signing time, after the environment approval was granted on the strength of the workflow file rather than the inputs. CodeRabbit's finding holds. Fix: regex-validate versions in `validate-inputs` (`^[0-9]+\.[0-9]+\.[0-9]+$`) and use the `$SOURCE`/`$VERSION` env vars already declared at `:85-87`. Same pattern at `release.yml:68-70, 76` (only the read-only token is exposed there).
4. **`release-qualification.yml:55, 67-71` — `environment: release-signing` + `ref: ${{ matrix.source }}`.** The environment's branch policy constrains the *workflow* ref, not the checkout; `validate-inputs` accepts any commit reachable after `fetch-depth: 0`, so the key can sign a never-reviewed branch commit and `release.yml:138` will tag it. Fix: `git merge-base --is-ancestor "$SOURCE_A" origin/main` in `validate-inputs`.
5. **`release.yml:106-119` — `publish-pypi` has `id-token: write` with no `environment:`.** PyPI trusted publishing binds to repo + workflow filename; `workflow_dispatch` can select any branch, so a modified `release.yml` on a feature branch can mint the token. Fix: put `publish-pypi` and `publish-github` behind a `release-publish` environment (required reviewer, branch policy `main`, no admin bypass) and register the trusted publisher with that environment name.

**Functional blockers**

6. **`release-qualification.yml:96` vs `script/v4-release` `_validate_source`** — `build` requires `git rev-parse "v<version>^{commit}"` to equal the source commit, so with no tag the signing job fails; if the operator pre-pushes `vA`/`vB` tags to get past it, `release.yml:138` 422s *after* the PyPI upload and the "reserved, never published" `vB` tag is public. Fix: `git tag "v$VERSION" "$SOURCE"` locally in the build job before invoking `build`, and add a tag/release pre-existence check before `publish-pypi`.
7. **`release-qualification.yml:125,129`, `release.yml:48,55,112,127`** — `actions/download-artifact@043fb46d…` does not exist in that repo (that SHA is `upload-artifact` v7.0.1), and `verify-approval` runs with `contents: read` only, whereas cross-run download needs `actions: read`. Both workflows fail at their first download; `test_signing_workflows.py` only string-matches YAML so cannot catch it. Fix: pin `download-artifact` to `37930b1c2abaa49bbe596cd826c3c89aef350131 # v7.0.0` (or v8.0.1 `3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c`), add `actions: read`.
8. **`release-qualification.yml:136-153`** — the per-OS step is `sha256sum -c` + evidence schema + `pip install wheel` + `--version`; it never runs `script/v4-release verify` (manifest signature / six-asset inventory), never exercises `install`, and candidate B is built and signed then used for nothing. `sha256sum` is absent on `macos-latest` (bot :141 holds; use `shasum -a 256 -c`). releasing.md's "verifies those exact artifacts on Linux, macOS, and Windows" overstates what runs. Also (bot release.yml:89) require the on-disk set under `release/` and `dist/` to equal the evidence inventory before publishing whole directories.

Minor: `docs/releasing.md:32` approval example shows strings where the workflow requires full evidence objects. What's right: no `pull_request_target`/`workflow_run` remain, least-privilege top-level `permissions`, `persist-credentials: false` everywhere, no checkout in `release.yml`, immutable artifacts rather than rebuilds between jobs, and the Windows bash paths are handled correctly.

**Comment by dsarno** [comment]
Addressed the eight review findings in c7c7357429b19464b8c63c41e90b6775b192f7cf, with one important distinction: these fixes make the release boundaries enforceable; they do **not** make the larger v4 qualification contract complete.

1. Promotion checks the canonical workflow/repository, dispatch event, main branch, successful run/attempt and every job **before** downloading artifacts. The approval binds run, attempt, workflow commit, both full candidate records and the complete evidence inventory.
2. The attestation repository is fixed to `dsarno/godot-ai-release-attestations`; only a bounded canonical JSON path at a full main-reachable commit is accepted.
3. Versions/source IDs are strictly validated; dispatcher and matrix values enter shell commands via quoted environment variables, never interpolated shell source.
4. A must be the workflow's reviewed main commit. B is an immediate child with exactly the two version edits and bundled README deletion. Builds have no signing key; the protected signer executes reviewed tooling against immutable source blobs.
5. Both publishing jobs use `release-publish`. The environment has a required reviewer, main-only policy and no admin bypass. The owner configured the matching environment-bound PyPI publisher and removed the old `(Any)` publisher.
6. Packaging tags exist only in disposable local clones. Public tag/release preflight happens before PyPI; matching partial uploads can resume, different bytes cannot overwrite an identity. Every B consumes the next patch number, so later patch bumps skip it; minor/major remain supported.
7. Download-artifact uses `37930b1c2abaa49bbe596cd826c3c89aef350131`, with explicit `actions: read`. Both workflows pass actionlint. Publication checks and edge cases now have executable Python tests, not only YAML string assertions.
8. Qualification verifies signatures, the exact six-asset release inventory, wheel/sdist identities and exact on-disk sets. The implemented Python producer retains dependencies, tests installed A and B wheels in matching source contexts, installs both wheel/sdist variants offline, and runs the documented signed installer for **both** candidates through a retained private index.

The example string-valued approval has been replaced with a command that emits the real full-object draft for independent human review. Publishing jobs check out only the reviewed workflow tooling; they neither execute candidate code nor rebuild release artifacts.

Also fixed the loaded-handler Update refusal and handler/dispatcher shutdown cycle. Built-in asynchronous continuations now retain an independent work ledger until they actually return; clearing request bookkeeping cannot authorize script replacement. Unknown custom handlers still fail closed. A busy preflight leaves the existing lifecycle live and retryable.

Validation: full expanded Python suite with Godot integration enabled **2,423 passed, four platform-only skips**; 135 focused release/index tests passed separately; full Ruff scope and actionlint passed. The complete provisional macOS/Python 3.14 package producer passed: **2,359 passed and 11 environment-inapplicable skips per installed A/B wheel**, both wheel/sdist variants installed offline, 79 retained distributions and both documented signed installs (279 A files; 278 B files). Fixture keys only; these are not final candidate approvals.

The manual Update after real tool use passed (plugin/backend 4.0.1, exact installed tree, retained backup, authenticated new backend, no crash report, clean Metal exit). The live suites on **both Godot 4.7 and 4.7.2** each passed 2,150 tests with zero failures and 11 conditional skips across 71 suites. Explicitly pinned exploratory stress passed 1,220 calls and 2/2 reloads with zero unexpected errors.

Release status is intentionally **not ready**: exact runtime, failpoint and locked-stress producers and reviewed numeric platform baselines remain missing. Qualification refuses **before build/sign** until those exist; a Python install row cannot authorize publication. The GLES3 full-suite texture-allocation warning also remains under investigation (the handler ObjectDB/resource leak is fixed; small no-Godot-AI controls have not reproduced the separate GPU warning). No final candidate was signed or published.

The durable eight-point disposition, test evidence and remaining gates are in [the release review checkpoint](https://github.com/hi-godot/godot-ai/blob/c7c7357429b19464b8c63c41e90b6775b192f7cf/docs/v4-release-review-followup.md). CI on this new revision is pending; earlier green checks do not qualify this commit.

File: .github/workflows/release-qualification.yml [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+          VERSION: ${{ matrix.version }}
+        shell: bash
+        run: |
+          set -euo pipefail
+          test -n "$RELEASE_SIGNING_KEY_PEM"
+          key_file="$(mktemp)"
+          trap 'rm -f "$key_file"' EXIT
+          printf '%s\n' "$RELEASE_SIGNING_KEY_PEM" > "$key_file"
+          python -m pip install --upgrade build==1.6.0
+          python script/v4-release build --output-dir candidate/release --private-key "$key_file" --channel stable --tag "v${{ matrix.version }}" --version "${{ matrix.version }}" --source-commit "${{ matrix.source }}"
```
[the remark this answers is not in the document]
Comment (dsarno) [reply .github/workflows/release-qualification.yml]: Fixed in c7c7357: strict canonical semantic versions/full source SHAs, and all dispatch/matrix values enter run scripts through quoted environment variables. No input expression is interpolated into signing shell source. Regression tests and actionlint pass.

File: .github/workflows/release-qualification.yml [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+          path: candidates/b
+      - uses: actions/setup-python@5fda3b95a4ea91299a34e894583c3862153e4b97 # v7.0.0
+        with:
+          python-version: "3.13"
+      - name: Verify artifact inventories and install exact wheels
+        shell: bash
+        run: |
+          set -euo pipefail
+          for candidate in candidates/*; do
+            (cd "$candidate" && sha256sum -c SHA256SUMS)
```
[the remark this answers is not in the document]
Comment (dsarno) [reply .github/workflows/release-qualification.yml]: Fixed in c7c7357: checksum and exact-inventory verification now use Python hashlib, not a platform-specific shell utility. The complete provisional macOS A/B package row passed, including both documented signed installs.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -14,12 +39,101 @@ env:
...
+          match = re.fullmatch(r"(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)", previous)
+          if not match: raise SystemExit("previous_version must be X.Y.Z")
+          records = {}
+          for path in Path(root).glob("*/evidence.json"):
+              record = json.loads(path.read_text(encoding="utf-8"))
+              if record.get("schema") != 1 or record.get("candidate") not in {"a", "b"}: raise SystemExit("invalid candidate evidence")
+              if record["candidate"] in records: raise SystemExit("duplicate candidate evidence")
+              for name, meta in record["files"].items():
+                  artifact = path.parent / name
+                  if not artifact.is_file() or hashlib.sha256(artifact.read_bytes()).hexdigest() != meta["sha256"] or artifact.stat().st_size != meta["size"]: raise SystemExit(f"candidate artifact mismatch: {name}")
```
[the remark this answers is not in the document]
Comment (dsarno) [reply .github/workflows/release.yml]: Fixed in c7c7357: candidate directories must equal the exact approved inventory, signatures and wheel/sdist identities are revalidated, and canonical successful workflow/run/attempt/all-job provenance is checked before cross-run downloads. Negative regressions cover extra/missing/changed files and invalid provenance. The complete release gate intentionally remains closed while runtime/failpoint/stress producers are absent.

File: .github/workflows/release-qualification.yml [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  qualify:
+    name: Verify immutable A/B bytes / ${{ matrix.os }}
+    needs: build
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+    steps:
+      - uses: actions/download-artifact@043fb46d1a93c77aae656e7c1c64a875d1fc6a0a # v7
```
[the remark this answers is not in the document]
Comment (dsarno) [reply .github/workflows/release-qualification.yml]: Fixed in c7c7357: every download-artifact use now pins 37930b1c2abaa49bbe596cd826c3c89aef350131 from the download repository, separately from upload-artifact. Both workflows pass actionlint.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -14,12 +39,101 @@ env:
...
-  publication-gate:
-    name: Refuse unqualified publication
+  verify-approval:
     runs-on: ubuntu-latest
+    outputs:
+      version: ${{ steps.identity.outputs.version }}
+      source: ${{ steps.identity.outputs.source }}
     steps:
-      - name: Require exact qualified-artifact promotion
+      - uses: actions/download-artifact@043fb46d1a93c77aae656e7c1c64a875d1fc6a0a # v7
```
[the remark this answers is not in the document]
Comment (dsarno) [reply .github/workflows/release.yml]: Fixed in c7c7357: corrected download-artifact SHA 37930b1c2abaa49bbe596cd826c3c89aef350131 and explicit actions: read for cross-run download jobs. Provenance is checked before each job downloads candidates.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -14,12 +39,101 @@ env:
...
+        env:
+          GH_TOKEN: ${{ github.token }}
         shell: bash
         run: |
-          echo "::error::v4 publication is disabled until immutable A/B qualification artifacts are bound to an approved promotion by run ID and digest."
-          exit 1
+          set -euo pipefail
+          [[ "${{ inputs.approval_commit }}" =~ ^[0-9a-f]{40}$ ]]
+          gh api "repos/${{ inputs.approval_repository }}/git/commits/${{ inputs.approval_commit }}" >/dev/null
+          gh api "repos/${{ inputs.approval_repository }}/contents/${{ inputs.approval_path }}?ref=${{ inputs.approval_commit }}" --jq .content | base64 --decode > approval.json
```
[the remark this answers is not in the document]
Comment (dsarno) [reply .github/workflows/release.yml]: Fixed in c7c7357: removed selectable approval_repository entirely; it is fixed to dsarno/godot-ai-release-attestations. The full commit and bounded relative JSON path are validated, and the commit must be reachable from attestation main. Invalid-path/noncanonical-approval regressions pass.

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -14,12 +39,101 @@ env:
...
+          if records["a"]["version"] != version or records["a"]["tag"] != "v" + version or not re.fullmatch(r"[0-9a-f]{40}", records["a"]["source"]): raise SystemExit("A identity does not match requested bump")
+          print("version=" + version)
+          print("source=" + records["a"]["source"])
+          PY
+      - uses: actions/upload-artifact@043fb46d1a93c77aae656e7c1c64a875d1fc6a0a # v7
+        with:
+          name: approved-v4-candidates
+          path: candidates/
+          retention-days: 90
+
```
[the remark this answers is not in the document]
Comment (dsarno) [reply .github/workflows/release.yml]: Fixed in c7c7357: GitHub tag/release preflight runs before PyPI writes. A pre-existing identity must match the approved source and exact asset subset; mismatched bytes are refused. Matching partial publication resumes without overwrites or skip-existing, with public PyPI bytes redownloaded and verified. Recovery cases have regression tests.

File: docs/releasing.md [hunk 8]
```diff
@@ -8,18 +8,40 @@ protocol. A temporary, signed v3-compatible capsule carries that same tree
...
+the already-qualified wheel and sdist first, then creates the tag and GitHub
+Release from A's already-qualified six assets. It never checks out or rebuilds
+the candidate. A duplicate PyPI version, changed artifact, missing B evidence,
+or mismatched approval fails closed.
+
+The approval file is JSON with this exact object shape (the `candidates` value
+is copied without changing any generated evidence fields):
+
+```json
+{"schema":1,"bump":"minor","previous_version":"4.0.0","candidates":{"a":"A evidence object","b":"B evidence object"}}
```
[the remark this answers is not in the document]
Comment (dsarno) [reply docs/releasing.md]: Fixed in c7c7357: replaced the string-valued example with the make-approval command, which emits canonical full A/B candidate records and the complete qualification inventory for independent human review. Tests verify the real full-object shape and reject altered bindings.

File: script/qualification_index.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            self.serve_artifact(head_only=True)
+
+        def serve_artifact(self, *, head_only):
+            path = urlsplit(self.path).path
+            if path in routes:
+                payload = routes[path]
+                self.send_response(200)
+                self.send_header("Content-Type", "text/html; charset=utf-8")
+            elif path.startswith(prefix + "/files/"):
+                name = path.removeprefix(prefix + "/files/")
```
[the remark this answers is not in the document]
Comment (dsarno) [reply script/qualification_index.py]: Fixed in 05c96d2: decode the URL path exactly once before the exact retained-inventory lookup. Real-network GET/HEAD tests now follow the generated link for 1.0+cpu, and encoded separators, double encoding and NUL suffixes remain refused. The index has 100% line coverage; a separate real uv install downloaded and imported a retained local-version wheel. Full local gauntlet passed (2,426 Python tests; 2,150 live Godot tests, all 71 suites). CI is running on this commit.

File: tests/unit/test_stormtest_support.py [hunk 10]
```diff
@@ -430,12 +434,146 @@ def test_repin_requires_one_new_session_with_same_pid_and_project(tmp_path):
...
+def test_exploratory_pin_records_identity_without_following_other_editor(tmp_path):
+    old = {"session_id": "old", "editor_pid": 42, "project_path": str(tmp_path)}
+    noise = {"session_id": "other", "editor_pid": 99, "project_path": str(tmp_path / "other")}
+    identity = support.pinned_session_identity([noise, old], "old")
+    # Windows project identities are deliberately case-normalized. Retain the
+    # exact PID/path assertion without requiring the original path's casing.
+    assert identity == {
+        "editor_pid": 42,
+        "project_path": os.path.normcase(str(tmp_path.resolve())),
+    }
```
[the remark this answers is not in the document]
Comment (dsarno) [reply tests/unit/test_stormtest_support.py]: Strengthened in 05c96d2: the fixture creates a deliberately mixed-case project directory. On Windows the initial session reports an uppercase path and the replacement reports lowercase; both must identify the same real directory, and the captured identity must equal the case-normalized resolved path. Exact PID binding and all refusal assertions remain intact. The full local gauntlet passed; the four Windows Python CI rows will exercise the platform-specific branch.

**Comment by dsarno** [comment]
Restart follow-up is pushed in 2589d41. The prior Windows rerun still timed out (31/32 jobs passed), so this is a tested fix, not another timeout increase:

- A minimal real-Godot regression proved that `--headless` is lost on editor restart. The harness now uses restart-forwarded display/audio driver options and checks distinct PIDs plus headless receipts in both the minimal restart and final-v3 migration.
- The bridge now saves v4's next-start enabled-plugin entry without enabling its scripts in the old process. Real-engine cases cover absent/already-listed entries, preserved companion plugins, save failure and malformed settings. Failure status is project-scoped.
- The existing predecessor-close wait now retries transient unverifiable process observations within its original deadline. Only positive dead/PID-reused proof transfers the nonce-bound lease; persistent uncertainty still refuses startup and preserves the old lease.
- CI emits per-test/progress output and retains replacement-editor diagnostics instead of leaving the Windows restart opaque.

Validation: **2,448 Python tests passed, four platform-inapplicable skips; 2,150 live Godot tests passed, zero failed, all 71 suites; owner-operated Update passed after real tool use; 1,223-call clean exploratory stress run had zero unexpected errors and survived both reloads.** Ruff, actionlint and all eleven architecture gates pass. The stress harness also gained typed AnyIO closed-stream classification, with tests proving errors remain unexpected outside the recorded reload window. Failed diagnostic runs are retained, not counted as passes.

All ten existing inline review threads are resolved. Full exact-candidate runtime/failpoint/stress qualification remains incomplete and the pre-signing guard stays closed; production signing/publishing permissions are unchanged. GLES texture shutdown diagnostics also remain under investigation. Windows CI on this new commit must pass before #957 is called resolved. The durable evidence and remaining gates are in `docs/v4-release-review-followup.md`.

**Comment by dsarno** [comment]
CI on 2589d41 has now completed successfully: **all 32 jobs passed**.

[Complete CI run](https://github.com/hi-godot/godot-ai/actions/runs/33715456363) · [Windows Godot job](https://github.com/hi-godot/godot-ai/actions/runs/33715456363/job/100523582312)

The expanded upgrade regression passed on all three platforms. Windows passed all 19 restart/bridge/update cases, including the formerly timing-out final-v3 migration in about 27.4 seconds, then passed the full Godot suites and reload/quit/export smokes. All twelve Python/OS matrix rows also passed.

The fresh owner-operated manual Update and local full gauntlet are recorded above. All ten existing inline review threads remain resolved. Merge approval is still pending; no final candidate was signed and nothing was published. The complete immutable-artifact/runtime, crash-recovery/failpoint, reviewed stress qualification and GLES texture investigation remain release work, with the pre-signing guard intentionally closed.

**Comment by dsarno** [comment]
Two overnight checkpoints are now tested:

- `293f58c` adds nine real activation-process kill/recovery rows at durable external barriers. They verify exact tree/journal states, startup refusal, refusal of repair while the editor-identity process is alive, and exact rollback after that process exits. These use a separate Python editor-identity process, not immutable-candidate Godot qualification. [All 32 CI jobs passed](https://github.com/hi-godot/godot-ai/actions/runs/33719622995).
- `91c4596` retains a minimal **native Godot project with no Godot AI code** that reproduces the two 349,524-byte GL texture warnings on macOS with official Godot 4.7.0 and 4.7.2. The combined-suite interaction is environment resource cleanup followed by immediate undo-history clearing. The same native sequence produces two warnings per environment; deferred-clear/retained-history controls do not. [Reproducer and control matrix](https://github.com/hi-godot/godot-ai/blob/91c4596/docs/verification/native-gles-environment-undo/README.md).

Fresh full validation before the diagnostic checkpoint: **2,457 Python tests passed, four platform skips; 2,150 live Godot tests passed, zero failed, all 71 suites; authenticated real read tools and normal editor quit passed.** Ruff, actionlint and all eleven architecture gates passed. No production code, test assertions, signing permissions or qualification gates changed. CI for the new diagnostic/documentation commit is separate from the completed code run above.

The native reproduction resolves the plugin-versus-engine isolation question, not internal engine cause, long-session impact or release disposition. No warnings were suppressed and no split-suite workaround was adopted. All ten inline review threads remain resolved. Exact-artifact runtime, complete external failpoint/recovery, and reviewed locked-stress qualification still remain; the pre-signing guard stays closed. No merge, production signing or publication occurred.

**Comment by dsarno** [comment]
Release-readiness checkpoint: the accumulated follow-up is committed as `84097d7`; the already reviewed #961 keepalive/logging fix is integrated separately on top.

Completed and verified:

- User-operated Update followed by normal editor Quit: fixture plugin/backend advanced to 4.0.1, durable migration completed, retained backup verified, old managed server stopped, no new Godot crash report. This is development-fixture evidence, not production candidate approval.
- Reload handoff now waits for the filesystem notification, preserves enabled state, and has real-engine persistence/timeout/ordering regressions.
- Qualification tooling covers additional external activation temporary-write, migration-completion and repair-claim boundaries, retained private HTTPS discovery/downloads, identity-bound resource sampling, and canonical baseline-only stress measurement. These remain partial producers, not completion of the full failpoint/runtime contract.
- Repeated stress testing exposed retained scratch-tab state. The harness now reloads and verifies an empty scratch scene before workers start. All failures remain retained; no error tolerance or acceptance threshold was weakened.
- Five corrected steady runs: 20,000 scheduled operations, 25,860 total calls, zero unexpected errors, exact cleanup each time. Local checkpoint: 2,697 Python tests passed / 4 skipped; 2,153 Godot tests passed / 11 skipped, plus reload and read probes. Ruff, actionlint and architecture gates pass.

Still blocking release approval:

- Editor RSS grew 29.75 MiB between the first and fifth post-60-second observations. The native GLES reproducer and 410 texture warnings remain unresolved; these are not a resource/no-growth pass.
- Full externally controlled recovery/lock matrix, immutable-candidate cross-platform runtime repetitions, and reviewed numeric storm/resource ceilings remain incomplete.
- Final source freeze, signing and independent attestation approvals have not happened. The qualification preflight still refuses before final signing; nothing was published.

Detailed evidence and scope limitations are retained in `docs/v4-release-review-followup.md`. The manual Update check no longer blocks this checkpoint.

The combined branch including #961 passed 2,699 Python tests / 4 skipped, plus the 2,153-test live Godot suite and real reload/read/normal-quit smoke. Hosted checks on the newly pushed revision remain a separate verification step; the earlier green checks are not being reused as proof of this revision.

**Comment by dsarno** [comment]
CI follow-up: all 32 jobs passed on head 92f4c87c5d81da5679ce416a9dcbb6fbd5af3a3b, including the expanded Godot/update regressions, Windows capture, reload/export smokes, and Python 3.11–3.14 across Linux/macOS/Windows. Run: https://github.com/hi-godot/godot-ai/actions/runs/33775412936 . This completes the local/hosted development checkpoint; it does not close the explicitly documented immutable-candidate qualification or resource/native-GLES release gates. Nothing was signed or published.

**Comment by dsarno** [comment]
Added the first immutable exact-candidate runtime qualification slice in 6c3cbb4. It binds signed A/B to the matching retained Python row, drives an unmodified A→B update in real Godot through the private HTTPS/index paths, verifies backend/client/tree/backup/transaction identities, and scans retained evidence for private capabilities. The aggregate validator now rejects missing, duplicate, unknown, or failed runtime/failpoint/stress case IDs; preflight intentionally remains closed until the other named producers exist.

Local validation: 2,675 Python tests passed (36 conditional/platform skips); focused release/security set 123 passed; explicit Godot-backed updater/migration/HTTPS/reload/restart set 38 passed; live Godot 4.7.2 corpus 2,140 passed, 0 failed, 24 headless/renderer skips across all 71 suites (2,164 total); Ruff, actionlint 1.7.7, shell syntax, bytecode compilation, whitespace, and all eleven architecture gates passed. Hosted CI is running.

**Comment by dsarno** [comment]
Engine identity gate added in e18167b. Qualification now checks exact executable size/SHA-256 before running even the version probe, uses the setup action's explicit executable path, rejects a differently labeled host OS, and independently validates retained engine identity during aggregation. The source-controlled manifest pins all six official Godot 4.7.0/4.7.2 desktop binaries. All six source archive digests matched official release metadata; their executable-entry sizes and hashes were verified independently.

Validation on the final tree: 2,695 Python tests passed, 36 conditional/platform skips; 142 focused release/security tests passed; 14 real-Godot driver/updater tests passed; 2,140 live headless Godot tests passed, zero failures, 24 skips, all 71 suites; normal editor/server shutdown passed. The new engine verifier has 100% line coverage. Ruff, actionlint, changed-file formatting, whitespace and all eleven architecture gates passed. The official macOS 4.7.2 executable also passed the real hash/version smoke.

A read-only configuration recheck confirmed main-only, owner-reviewed signing/publishing environments, signing-secret presence (metadata only), and a public attestation repository with Actions disabled and only dsarno listed as a writer. No secret values were read and no settings were changed. This is not a key-fingerprint or actual release-credential scope proof.

The preceding ef2d53d checkpoint passed all 32 hosted CI jobs: https://github.com/hi-godot/godot-ai/actions/runs/33793021903 . This new commit requires its own CI run. No production plugin/server code, Windows test timing, signing permissions or qualification thresholds changed. Exact-candidate runtime/failpoint/stress qualification remains incomplete, the pre-signing guard stays closed, and nothing was merged, signed or published.
